### DISCUSSION TITLE
Is there any way to create user session from admin api?

### DISCUSSION THREAD

**Discussion by nxy7** [ory/kratos#3230] [Q&A] [answered]
Hello, I'm creating some integration tests for my app and to simulate user requests I'd like to somehow get ory_kratos_session cookies using admin API. My understanding is that right now session cookies are distributed only by self-service flows, but I figured I'd ask about it here before making feature request.
My current workflow consists of creating user by post request at /admin/identities. I store ID from response and ideally I'd like to make another request with that ID to create session.
So is there currently any way of doing that?

**Comment by vinckr** [comment, marked as answer]
Hey @nxy7 
for the browser flows you do need a browser to get a session. 
For testing I would suggest something like [Playwright](https://playwright.dev/), that is what we use internally to test these things.

See in our docs: https://github.com/ory/docs/blob/master/tests/playwright/protect-page-login.spec.ts

**Reply by nxy7** [reply]
I'd like to avoid bringing frontend into integration tests so I'll look for some workarounds. Thanks for confirming that there's no 'easy' way of doing that right now.

**Reply by vinckr** [reply]
Makes sense yea. Let me know if you find a better solution / workaround!

**Reply by nxy7** [reply]
Just to leave some comment I was thinking about using API login flow for my tests, but as it is so much different that normal flow of my app (and I think session tokens returned for API flows and cookie sessions are not the same thing) I'll just mock out kratos (after all I only need 2/3 endpoints). I think that's the best approach here.

**Reply by vinckr** [reply]
For a web application always use the browser API!  
It is potentially insecure and dangerous to use the native flows in browser applications.

>session tokens returned for API flows and cookie sessions are not the same thing

Yes they are not the same thing, the cookies have built-in security from browsers, whereas with native applications you use the device keychain or something similar.

**Reply by nxy7** [reply]
I wasn't going to use flow API version in app, just in my tests. I've wanted to simulate user in tests by getting session token and making requests from integration tests with ORY_KRATOS_COOKIE={SESSION_TOKEN} but as You've confirmed they are not the same.
I wasn't planning on using this mechanism in real app.
Anyway mocking will both be faster (tests duration) and easier. Thanks for clearing things up :-)
